### PR TITLE
Dockstore import

### PR DESCRIPTION
**Pull request by MattPutnam**
To see it in action, go to a URL such as
```
http://localhost:3000/#import/dockstore/github.com/DataBiosphere/topmed-workflows/Functional_Equivalence:1.0.0
```
There's a lot of placeholder stuff, but this PR is already massive so I'm trying to find a good place to cut this off.

### CODE REVIEW

File: src/components/Collapse.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    return div(_.omit(this.props, ['defaultHidden', 'title', 'titleExpand']), [
+      div({ style: { marginBottom: '0.5rem' } }, [
+        div(
+          {
+            style: { display: 'flex', cursor: 'pointer', alignItems: 'center' },
+            onClick: () => this.setState({ visible: !visible })
+          },
+          [
+            span({ style: { color: Style.colors.secondary } }, title),
+            icon(visible ? 'angle down' : 'angle left', { style: { marginLeft: '0.25rem' } })
```
Comment (MattPutnam) [inline src/components/Collapse.js:27]: Open question: do we like this design for the collapse icon? Seems more natural for the icon to go on the left, but then in FC we ran into the issue where that pushed the title to the right and it didn't line up with the form.

File: src/components/TopBar.js [hunk 2]
```diff
@@ -32,6 +32,7 @@ export class TopBar extends Component {
       div(
         {
           style: {
+            zIndex: 7, // CodeMirror uses zIndexes, this seems to be enough.
```
Comment (MattPutnam) [inline src/components/TopBar.js:35]: Please tell me there's a better way to do this.

File: src/pages/Import.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+          }, [
+            icon('error'),
+            JSON.parse(importError).message
+          ]) : null
+      ]
+    )
+  }
+
+  import = () => {
+    const { selectedWorkspace: { data: { workspace: { namespace, name } } } } = this.state
```
Comment (MattPutnam) [inline src/pages/Import.js]: Having to unpack `data` makes me feel like I'm using the Select wrong.

File: src/pages/Import.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            disabled: !selectedWorkspace,
+            onClick: () => this.import()
+          },
+          'Import'),
+        importError ?
+          div({
+            style: { marginTop: '1rem', color: Style.colors.error }
+          }, [
+            icon('error'),
+            JSON.parse(importError).message
```
Comment (MattPutnam) [inline src/pages/Import.js:118]: I'd love to see us come up with a general solution to error responses. @dmohs mentioned that his startup handles all errors with something like FC's notification bar.

File: src/components/TopBar.js (hunk 2 above)
Comment (dmohs) [reply src/components/TopBar.js:35]: I would imagine the other option is to override CodeMirror's styles with sniper CSS.

File: src/components/WDLViewer.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+})
+
+
+export default class WDLViewer extends Component {
+  render() {
+    return div({ ref: elem => this.container = elem, style: { border: Style.standardLine } })
+  }
+
+  componentDidMount() {
+    CodeMirror(this.container, { mode: 'wdl', value: this.props.wdl, lineNumbers: true })
```
Comment (dmohs) [inline src/components/WDLViewer.js:41]: Do you know if this needs any cleanup when unmounting?

File: src/components/Collapse.js (hunk 1 above)
Comment (dmohs) [reply src/components/Collapse.js:27]: Personally, I think it looks great there.

File: src/pages/Import.js (hunk 4 above)
Comment (dmohs) [reply src/pages/Import.js:118]: I do, however, really like the placement of the error messages right under the submit button in this case. Conceptually, they seem to appear in the right spot.

**Review by dmohs — APPROVED** [review]
Looks really good.

File: src/pages/Import.js (hunk 3 above)
Comment (panentheos) [reply src/pages/Import.js]: The Select API only requires `value` and `label` in options, where `value` is the underlying data being selected. That would be the workspace, in this case. The `onChange` callback gives you back the whole option, so you could just destructure out `value` at that point, and store just that.

File: src/libs/ajax.js [hunk 6]
```diff
@@ -147,6 +146,20 @@ export const Rawls = {
       entity: (type) => {
         return fetchRawls(`${root}/entities/${type}`, authOpts())
           .then(parseJson)
+      },
+
+      methodConfigs: () => {
```
Comment (panentheos) [inline src/libs/ajax.js]: What is the benefit of this function call?

File: src/pages/Import.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+class Importer extends Component {
+  render() {
+    const { source } = this.props
+
+    return h(Fragment, [
+      h(TopBar, { title: 'Import' }),
+      Utils.cond(
+        [source === 'dockstore', this.renderDockstore],
+        () => `Unknown source '${source}'`
```
Comment (panentheos) [inline src/pages/Import.js:162]: What is this indirection for? Do we know for sure that we will have multiple import workflows on this page?

File: src/pages/Import.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        h(Select, {
+          clearable: false,
+          searchable: true,
+          disabled: !workspaces,
+          placeholder: workspaces ? 'Select a workspace' : 'Loading workspaces...',
+          value: selectedWorkspace,
+          onChange: selectedWorkspace => this.setState({ selectedWorkspace }),
+          options: _.map(workspaces, workspace => {
+            const { name } = workspace.workspace
+            return { value: name, label: name, data: workspace }
```
Comment (panentheos) [inline src/pages/Import.js]: `value` should be unique among the options. This should probably be the whole workspace, not just the name.

**Review by panentheos — APPROVED** [review]
Looks good (see suggestions about Select usage)

File: src/components/WDLViewer.js (hunk 5 above)
Comment (MattPutnam) [reply src/components/WDLViewer.js:41]: Doesn't seem like it. FireCloud doesn't do anything, and I don't see anything in the codemirror documentation.
